api/src/transactions.rs

```diff
@@ -360,7 +368,74 @@ impl TransactionsApi {
             return Err(api_disabled("Simulate transaction"));
         }
         let ledger_info = self.context.get_latest_ledger_info()?;
-        let signed_transaction = self.get_signed_transaction(&ledger_info, data)?;
+        let mut signed_transaction = self.get_signed_transaction(&ledger_info, data)?;
+
+        let estimated_gas_unit_price = if estimate_gas_unit_price.0.unwrap_or_default() {
+            Some(self.context.estimate_gas_price(&ledger_info)?)
+        } else {
+            None
+        };
+
+        // If estimate max gas amount is provided, we will just make it the maximum value
+        let estimated_max_gas_amount = if estimate_max_gas_amount.0.unwrap_or_default() {
+            // Retrieve max possible gas units
+            let gas_params = self.context.get_gas_schedule(&ledger_info)?;
+            let max_number_of_gas_units = u64::from(gas_params.txn.maximum_number_of_gas_units);
+
+            // Retrieve account balance to determine max gas available
+            let account_state = self
+                .context
+                .get_account_state(
+                    signed_transaction.sender(),
+                    ledger_info.version(),
+                    &ledger_info,
+                )?
+                .ok_or_else(|| {
+                    SubmitTransactionError::bad_request_with_code(
+                        "Account not found",
+                        AptosErrorCode::InvalidInput,
+                        &ledger_info,
+                    )
+                })?;
+            let coin_store: CoinStoreResource = account_state
+                .get_coin_store_resource()
+                .and_then(|inner| inner.ok_or_else(|| anyhow!("No coin store found!")))
+                .map_err(|err| {
+                    SubmitTransactionError::internal_with_code(
+                        format!("Failed to get coin store resource {}", err),
+                        AptosErrorCode::InternalError,
+                        &ledger_info,
+                    )
+                })?;
+
+            let gas_unit_price =
+                estimated_gas_unit_price.unwrap_or_else(|| signed_transaction.gas_unit_price());
+
+            // With 0 gas price, we set it to max gas units, since we can't divide by 0
+            let max_account_gas_units = if gas_unit_price == 0 {
+                coin_store.coin()
+            } else {
+                coin_store.coin() / gas_unit_price
```

Comment:
Can we not do division and instead compare using multiplication (+ even casting to u128 if needed)? Multiplication can lead to off by 1 errors due to rounding

Reply:
This handles the following case:

I have a global max amount of gas units of 3.

The gas price is 10.

If I have 20-29 coins, I can only pay for gas 2 gas units (e.g. 29 / 10 rounded down).

I compare 2 to 3, and I have the minimum that is supposed to be 2 and not 3.

If I have 39 coins, I can only pay for 3 gas units (e.g. 39 / 10 rounded down).

I compare 3 to 3 and I have the minimum of that which is 3.

If I have 40 coins, I can only pay for 4 gas units, but the max number I can have is 3.

Min(4, 3) = 3